### PR TITLE
integrate neural network into Snow.jl architecture (Rebased)

### PR DESCRIPTION
**Pull request by kmdeck**
## Content

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/standalone/Snow/Snow.jl [hunk 1]
```diff
@@ -203,12 +259,12 @@ such that SWE cannot become negative and U cannot become unphysical. The
 clipped values are what are actually applied as boundary fluxes, and are stored in
 `applied_` fluxes.
 """
-auxiliary_vars(::SnowModel) = (
+auxiliary_vars(m::SnowModel) = (
```
Comment (kmdeck) [inline src/standalone/Snow/Snow.jl:262]: this is nitpicky (sorry!), but we can just to ::SnowModel since `m` is not needed in the function. it will just dispatch off of type

File: src/standalone/Snow/Snow.jl [hunk 2]
```diff
@@ -316,8 +380,12 @@ end
...
 snow water equivalent S will not become negative in a timestep Δt.
 """
-function clip_total_snow_water_flux(S, total_water_flux, Δt)
-    if S - total_water_flux * Δt < 0
+function clip_total_snow_water_flux(
+    S::FT,
+    total_water_flux::FT,
+    Δt::FT,
+) where {FT}
+    if S - total_water_flux * Δt < eps(FT)
```
Comment (kmdeck) [inline src/standalone/Snow/Snow.jl:388]: Were these changes required?

File: src/standalone/Snow/snow_parameterizations.jl [hunk 3]
```diff
@@ -23,27 +23,38 @@ function snow_cover_fraction(x::FT; α = FT(1e-3))::FT where {FT}
...
 
+"""
+    snow_depth(model::AbstractDensityModel, Y, p, params)
+
+A default function which returns the snow depth through the scaling of SWE by the snow density.
+This can be extended for additional types of parameterizations.
+"""
+function snow_depth(density::AbstractDensityModel{FT}, Y, p, params) where {FT}
+    ρ_l = FT(LP.ρ_cloud_liq(params.earth_param_set))
+    return ρ_l .* Y.snow.S ./ p.snow.ρ_snow
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:34]: can replace with @. ρ_l * Y.snow.S / p.snow.ρ_snow

File: src/standalone/Snow/snow_parameterizations.jl [hunk 4]
```diff
@@ -23,27 +23,38 @@ function snow_cover_fraction(x::FT; α = FT(1e-3))::FT where {FT}
...
-height of the atmosphere measurement location (z_atmos > z_land always).
-
-This assumes that the surface elevation is zero.
+In order to compute surface fluxes, this cannot be large than the 
+height of the atmosphere measurement location (z_atmos > z_land always). 
 """
-function ClimaLand.surface_height(model::SnowModel{FT}, Y, p) where {FT}
-    return FT(0)
+function ClimaLand.surface_height(model::SnowModel, Y, p)
+    return snow_depth(model.parameters.density, Y, p, model.parameters)
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:57]: we should revert this change and change in doc string

File: src/standalone/Snow/snow_parameterizations.jl [hunk 5]
```diff
@@ -165,23 +163,15 @@ Computes the thermal conductivity, given the density
...
     ρ_snow::FT,
     parameters::SnowParameters{FT},
 ) where {FT}
     _κ_air = FT(LP.K_therm(parameters.earth_param_set))
-    _ρ_ice = FT(LP.ρ_cloud_ice(parameters.earth_param_set))
     κ_ice = parameters.κ_ice
     return _κ_air +
-           (FT(0.07) * (ρ_snow / _ρ_ice) + FT(0.93) * (ρ_snow / _ρ_ice)^2) *
-           (κ_ice - _κ_air)
+           (FT(7.75e-5) * ρ_snow + FT(1.105e-6) * ρ_snow^2) * (κ_ice - _κ_air)
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:174]: Let's also revert these changes and change to doc string

File: src/standalone/Snow/Snow.jl [hunk 6]
```diff
@@ -272,8 +328,15 @@ function ClimaLand.make_update_aux(model::SnowModel{FT}) where {FT}
...
 
-        @. p.snow.water_runoff =
-            compute_water_runoff(Y.snow.S, p.snow.q_l, p.snow.T, parameters)
+        p.snow.water_runoff .=
+            compute_water_runoff.(
+                Y.snow.S,
+                p.snow.q_l,
+                p.snow.T,
+                p.snow.ρ_snow,
+                snow_depth(model.parameters.density, Y, p, parameters),
```
Comment (kmdeck) [inline src/standalone/Snow/Snow.jl:337]: Question: can we always use snow_depth = SWE*rho_l/rho_snow? We just updated density above on line 319 using the density model correctly.

File: src/standalone/Snow/snow_parameterizations.jl [hunk 7]
```diff
@@ -121,19 +132,6 @@ as the bulk temperature T.
...
 
 
-"""
-    snow_depth(SWE::FT, ρ_snow::FT, ρ_l::FT) where {FT}
-
-Returns the snow depth given SWE, snow density ρ_snow, and
-the density of liquid water ρ_l.
-
-"""
-function snow_depth(SWE::FT, ρ_snow::FT, ρ_l::FT)::FT where {FT}
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:131]: see above comment - can we keep this definition?

File: src/standalone/Snow/snow_parameterizations.jl [hunk 8]
```diff
@@ -203,18 +193,33 @@ function snow_bulk_temperature(
...
     _LH_f0 = FT(LP.LH_f0(parameters.earth_param_set))
     cp_s = specific_heat_capacity(q_l, parameters)
-    _T_freeze = FT(LP.T_freeze(parameters.earth_param_set))
     _ρcD_g = parameters.ρcD_g
     return _T_ref +
            (U + (1 - q_l) * _LH_f0 * _ρ_l * SWE) / (_ρ_l * SWE * cp_s + _ρcD_g)
 
 end
 
+"""
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:206]: Does this make sense with the constant density model? In that case, we have rho_snow = constant, but here we return a different value. 

We also do not divide by SWE here, so what NaN arises? Should the restriction be placed on z instead?

File: src/standalone/Snow/snow_parameterizations.jl [hunk 9]
```diff
@@ -226,27 +231,25 @@ function snow_liquid_mass_fraction(
     SWE::FT,
     parameters::SnowParameters{FT},
 ) where {FT}
-    _ρ_i = FT(LP.ρ_cloud_ice(parameters.earth_param_set))
+    #if SWE < eps(FT) return FT(1) end #do this to speed up computation or not, compared to the max() in the final line?
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:234]: I think what we have is OK

File: src/standalone/Snow/snow_parameterizations.jl [hunk 10]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+end
+
+"""
+    newsnow_density(air_temp::FT)::FT where {FT}
+Estimates the density of newly fallen snow as a function of air temperature, according to the Snow17 model.
+Used in computing the time derivative of snow depth under the Anderson1976 density parameterization.
+"""
+function newsnow_density(air_temp::FT)::FT where {FT}
+    if air_temp < -15
+        return FT(0.05)
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:471]: what units are these? Is 0.05 kg/m^3 reasonable?

File: src/standalone/Snow/snow_parameterizations.jl [hunk 11]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+    end
+end
+
+"""
+    newsnow_temp(air_temp::FT)::FT where {FT}
+Estimates the temperature of newly fallen snow as a function of air temperature, according to the Snow17 model.
+Used in computing the time derivative of snow depth under the Anderson1976 density parameterization.
+"""
+function newsnow_temp(air_temp::FT)::FT where {FT}
+    return (air_temp > 0) ? FT(0) : air_temp
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:483]: Our temps are always in Kelvin - so I think this has to be updated

File: src/standalone/Snow/snow_parameterizations.jl [hunk 12]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+"""
+    newsnow_density(air_temp::FT)::FT where {FT}
+Estimates the density of newly fallen snow as a function of air temperature, according to the Snow17 model.
+Used in computing the time derivative of snow depth under the Anderson1976 density parameterization.
+"""
+function newsnow_density(air_temp::FT)::FT where {FT}
+    if air_temp < -15
+        return FT(0.05)
+    else
+        return FT(0.05) + FT(0.0017) * FT((air_temp + FT(15.0))^1.5)
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:473]: temps are in Kelvin

File: src/standalone/Snow/snow_parameterizations.jl [hunk 13]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+    return (air_temp > 0) ? FT(0) : air_temp
+end
+
+"""
+    dzdt(density::Anderson1976{FT}, model::SnowModel{FT}, Y, p, t) where {FT}
+Returns the change in snow depth (rate) given the current model state and the `Anderson1976`
+density paramterization, which estimates contributions from new snowfall, as well as compaction
+effects from the existing and newly fallen snow.
+"""
+function dzdt(density::Anderson1976{FT}, model::SnowModel{FT}, Y, p) where {FT}
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:492]: It would be fantastic to have unit tests for these or at least to have a simulation running with this parameterization.

File: src/standalone/Snow/snow_parameterizations.jl [hunk 14]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+"""
+    clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT
+A helper function which clips the tendency of Z such that
+its behavior is consistent with that of S.
+"""
+function clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT where {FT}
+    if (S + dSdt * Δt) <= eps(FT)
+        return -Z / Δt
+    elseif (Z + dZdt * Δt) < (S + dSdt * Δt)
+        #do we need to do something about setting q_l = 1 here?
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:533]: q_l is determined independently, so we shouldnt be changing it here.

File: src/standalone/Snow/snow_parameterizations.jl [hunk 15]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+    ρ_est = compact_density.(W_use, sliq, ρ_ice, Δt_t, Tsnow, Ref(density))
+    dz_compaction = (W_ice ./ ρ_est) .- Y.snow.Z
+
+    return (dz_compaction .+ Δz) ./ model.parameters.Δt
+end
+
+"""
+    clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT
+A helper function which clips the tendency of Z such that
+its behavior is consistent with that of S.
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:527]: "A helper function which clips the tendency of Z such that
its behavior is consistent with that of S: if all snow melts within a timestep, we clip the tendency of S so that it does not become negative, and here we also clip the tendency of Z so that depth does not become negative. Additionally, if the tendencies of Z and S are such that we would encounter Z < S (rho_snow > rho_liq), we also clip the tendency.

File: src/standalone/Snow/snow_parameterizations.jl [hunk 16]
```diff
@@ -372,3 +376,178 @@ function energy_from_T_and_swe(S::FT, T::FT, parameters) where {FT}
...
+"""
+function clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT where {FT}
+    if (S + dSdt * Δt) <= eps(FT)
+        return -Z / Δt
+    elseif (Z + dZdt * Δt) < (S + dSdt * Δt)
+        #do we need to do something about setting q_l = 1 here?
+        return (S - Z) / Δt + dSdt
```
Comment (kmdeck) [inline src/standalone/Snow/snow_parameterizations.jl:534]: Could you describe where this comes from?

File: test/standalone/Snow/parameterizations.jl [hunk 17]
```diff
@@ -67,14 +72,13 @@ for FT in (Float32, Float64)
         )
 
         SWE = cat(FT.(rand(100) .+ 0.2), FT(0), dims = 1)
-        z = snow_depth.(SWE, ρ_snow, _ρ_l)
-        @test all(z .== SWE * _ρ_l ./ ρ_snow)
+        z = SWE * _ρ_l ./ ρ_snow
+        #@test all(z .== SWE * _ρ_l ./ ρ_snow)
```
Comment (kmdeck) [inline test/standalone/Snow/parameterizations.jl:76]: uncomment or remove?

File: test/standalone/Snow/parameterizations.jl [hunk 18]
```diff
@@ -67,14 +72,13 @@ for FT in (Float32, Float64)
...
         @test specific_heat_capacity(FT(1.0), parameters) == _cp_l
         @test specific_heat_capacity(FT(0.0), parameters) == _cp_i
         @test snow_thermal_conductivity(ρ_snow, parameters) ==
               κ_air +
-              (FT(0.07) * (ρ_snow / _ρ_i) + FT(0.93) * (ρ_snow / _ρ_i)^2) *
-              (κ_ice - κ_air)
+              (FT(7.75e-5) * ρ_snow + FT(1.105e-6) * ρ_snow^2) * (κ_ice - κ_air)
```
Comment (kmdeck) [inline test/standalone/Snow/parameterizations.jl:81]: should change back

File: src/integrated/soil_snow_model.jl [hunk 19]
```diff
@@ -253,7 +253,7 @@ function update_soil_snow_ground_heat_flux!(
     κ_soil = ClimaLand.Domains.top_center_to_surface(p.soil.κ)
 
     # Depth of snow and soil layers interacting thermally at interface
-    Δz_snow = p.snow.z # Snow depth
+    Δz_snow = Snow.snow_depth(snow_params.density, Y, p, snow_params) #Snow depth
```
Comment (kmdeck) [inline src/integrated/soil_snow_model.jl:256]: I am wondering why we cannot do the formula Y.snow.S * p.snow.rho_snow / rho_liq here

File: src/integrated/soil_snow_model.jl [hunk 20]
```diff
@@ -387,22 +387,19 @@ function soil_boundary_fluxes!(
...
     p.soil.R_n .= net_radiation(bc.radiation, soil, Y, p, t)
-    # influx = maximum possible rate of infiltration given precip, snowmelt, evaporation/condensation
-    # but if this exceeds infiltration capacity of the soil, runoff will
-    # be generated.
-    # Use top_bc.water as temporary storage to avoid allocation
-    influx = p.soil.top_bc.water
-    @. influx =
-        p.drivers.P_liq +
-        p.snow.water_runoff * p.snow.snow_cover_fraction +
-        p.excess_water_flux +
```
Comment (kmdeck) [inline src/integrated/soil_snow_model.jl:398]: we need to keep these changes.

File: src/integrated/soil_snow_model.jl [hunk 21]
```diff
@@ -387,22 +387,19 @@ function soil_boundary_fluxes!(
...
+        p.excess_water_flux,
+        Y,
+        t,
+        soil,
+    )
+    @. p.soil.top_bc.water =
+        p.soil.infiltration +
         (1 - p.snow.snow_cover_fraction) *
         p.soil.turbulent_fluxes.vapor_flux_liq
-    # The update_runoff! function computes how much actually infiltrates
```
Comment (kmdeck) [inline src/integrated/soil_snow_model.jl:401]: like wise here

File: ext/neural_snow/NeuralSnow.jl [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+            ) .- Y.snow.Qrel_avg
+        )
+    @. dY.snow.u_avg =
+        density.α / model.parameters.Δt * (p.drivers.u - Y.snow.u_avg)
+end
+
+end
+
+#Todo: Write tests for these functions
+#Todo: address Anderson1976 sensitivity for small precision snowpacks (z = 0 when SWE > 0 or z > 0 when SWE = 0, but on the order of 1e-10 < eps(Float32), doesn't make NaN at present though)
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:213]: I propose we move Anderson1976 from this PR if you do not feel up to writing tests for it and want to merge something closer to what you have already. And then you can immediately open another PR adding in the Anderson parameterization code (with unit tests, the correct units, etc)

File: ext/neural_snow/NeuralSnow.jl [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+    density_prog_types,
+    density_prog_names,
+    update_density!,
+    update_density_prog!
+import ClimaLand.Parameters as LP
+using Thermodynamics
+
+using DataFrames, Dates, CSV, HTTP, Flux, cuDNN, StatsBase, BSON
+ModelTools = Base.get_extension(ClimaLand, :NeuralSnowExt).ModelTools
+#^^is this correct format since this is within the extension itself, or do i just use include(./ModelTools); using ModelTools; here?
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:17]: I don't know the answer to this, but this seems more in line with what I do in other places:
include("./ModelTools.jl"); using ModelTools;

Comment (kmdeck) [reply ext/neural_snow/NeuralSnow.jl:17]: e.g. here: https://github.com/CliMA/ClimaLand.jl/blob/ccc6009016f80f9ce89c42eaaf61a9fe8e9ff410/src/ClimaLand.jl#L7-L8

File: ext/neural_snow/NeuralSnow.jl [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+
+export NeuralDepthModel, snow_depth, update_density_prog!
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:19]: I do not think that we need to export these: , snow_depth, update_density_prog!

File: ext/neural_snow/NeuralSnow.jl [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+    z_model::Flux.Chain
+    α::FT
+end
+
+"""
+    get_znetwork(; download_link = "https://caltech.box.com/shared/static/ay7cv0rhuiytrqbongpeq2y7m3cimhm4.bson")
+A default function for returning the snow-depth neural network from the paper.
+"""
+#This is the only place that creates the necessity of BSON for the neural extension and the make_model() from ModelTools. Keep or get rid of?
+#Can I change this to something CliMA already includes? is BSON vs JLD2 or anything else better? An artifact (can they store BSON/.jld2 files for direct usage without those packages?)?
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:36]: I would look into what Flux recommends for storing models. In the past I have had trouble reading (old) Flux models saved to BSON if Flux has updated.

File: ext/neural_snow/NeuralSnow.jl [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+function get_znetwork(;
+    download_link = "https://caltech.box.com/shared/static/ay7cv0rhuiytrqbongpeq2y7m3cimhm4.bson",
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:38]: Let's expose this in src/Artifacts.jl.

Even something as simple as 
neural_snow_znetwork_link() = "https://caltech.box.com/shared/static/ay7cv0rhuiytrqbongpeq2y7m3cimhm4.bson"

e.g. here:https://github.com/CliMA/ClimaLand.jl/blob/ccc6009016f80f9ce89c42eaaf61a9fe8e9ff410/src/Artifacts.jl#L233-L242

File: ext/neural_snow/NeuralSnow.jl [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+    z_idx = 1
+    p_idx = 7
+    nfeatures = 7
+    zmodel = ModelTools.make_model(nfeatures, 4, z_idx, p_idx)
+    zmodel_state = BSON.load(IOBuffer(HTTP.get(download_link).body))[:zstate]
+    Flux.loadmodel!(zmodel, zmodel_state)
+    ModelTools.settimescale!(zmodel, 86400.0)
+    return zmodel
+
+    #should I write this function to give the ability to also load from local file too?
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:49]: You could have this function take the path, and compute the path in current sims from the suggested function above

File: ext/neural_snow/NeuralSnow.jl [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+
+"""
+    snow_depth(m::NeuralDepthModel, Y, p, params)
+
+An extension of the `snow_depth` function to the NeuralDepthModel density parameterization, which includes the prognostic 
+depth variable and thus does not need to derive snow depth from SWE and density.
+This is sufficient to enable dynamics of the auxillary variable `ρ_snow` without extension of update_density!, and avoids
+redundant computations in the computation of runoff.
+"""
+ClimaLand.Snow.snow_depth(m::NeuralDepthModel, Y, p, params) = Y.snow.Z
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:115]: we need to import this also

File: ext/neural_snow/NeuralSnow.jl [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+    model,
+    z::FT,
+    swe::FT,
+    P::FT,
+    T::FT,
+    R::FT,
+    qrel::FT,
+    u::FT,
+)::FT where {FT}
+    return model([z, swe, qrel, R, u, T, P])[1]
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:133]: Could you explain why you need the [1] index?

File: ext/neural_snow/NeuralSnow.jl [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+        Y.snow.u_avg,
+    )
+end
+
+"""
+    clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT
+A helper function which clips the tendency of Z such that
+its behavior is consistent with that of S.
+"""
+function clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT where {FT}
```
Comment (kmdeck) [inline ext/neural_snow/NeuralSnow.jl:160]: Does this duplicate the function you wrote in src/standalone/Snow? If so, we can use Snow.clip_dZdt instead of redefining here.

File: docs/Project.toml [hunk 31]
```diff
@@ -1,5 +1,7 @@
 [deps]
+About = "69d22d85-9f48-4c46-bbbe-7ad8341ff72a"
```
[the remark this answers is not in the document]
Comment (a-charbon) [reply docs/Project.toml:2]: sounds good!

File: src/standalone/Snow/Snow.jl [hunk 32]
```diff
@@ -42,6 +42,49 @@ and is used as a bulk snow model.
...
+end
+
+"""
+    Anderson1976{FT <: AbstractFloat} <: AbstractDensityModel{FT}
+Establishes the density parameterization where snow density
+compacts according to the seminal works of Anderson in the 1970s (and specifically the
+numerical implementation documented in the Snow17 model).
+This uses the default constants defined in the Snow17 documentation, but allows for redefinition of the model constants.
+These include:
+- c1: fractional rate of density increase from compaction at 0ᵒC (0.026 cm⁻¹ hr⁻¹; cm of water-equivalent load)
```
[the remark this answers is not in the document]
Comment (a-charbon) [reply src/standalone/Snow/Snow.jl:69]: No, this is handled a bit sneakily through the addition of the "100" prefactor -
Snow17 includes "0.1 * W_i" with W_i in millimeters (to create correct units with the cm/hr constants), but i've put "100 * W_i" as we provide W_i in meters, and unit consistency is maintained. This also allows minimal changes between the Snow17 and our formulation of the same function, and creates the most consistency between the construction of an Anderson1976 model and constructing a Snow17 run (as well as consistency with the literature).

I've added a comment to the comapct_density() function and function descriptions to make this more clear.

File: src/standalone/Snow/Snow.jl [hunk 33]
```diff
@@ -42,6 +42,49 @@ and is used as a bulk snow model.
...
+- c2: compression constant estimated by Kojima in 1967 (21 cm³/g)
+- c3: fractional settling rate (destructive metamorphism) at 0ᵒC for densities less than the critical density ρ_d (0.005 hr⁻¹)
+- c4: constant for adjusting fractional settling due to temperature (0.1 ᵒC⁻¹)
+- c5: scaling of the settling rate when no water is present in the snowpack (sources vary on whether this factor should be 0 or 1, default is set to 0)
+- cx: destructive metamorphism decay factor for densities greater than the critical density ρ_d (default value is 23)
```
[the remark this answers is not in the document]
Comment (a-charbon) [reply src/standalone/Snow/Snow.jl:74]: unitless, "(unitless)" has now been added.

File: src/standalone/Snow/Snow.jl [hunk 34]
```diff
@@ -42,6 +42,49 @@ and is used as a bulk snow model.
...
+"""
+    Anderson1976{FT <: AbstractFloat} <: AbstractDensityModel{FT}
+Establishes the density parameterization where snow density
+compacts according to the seminal works of Anderson in the 1970s (and specifically the
+numerical implementation documented in the Snow17 model).
+This uses the default constants defined in the Snow17 documentation, but allows for redefinition of the model constants.
+These include:
+- c1: fractional rate of density increase from compaction at 0ᵒC (0.026 cm⁻¹ hr⁻¹; cm of water-equivalent load)
+- c2: compression constant estimated by Kojima in 1967 (21 cm³/g)
+- c3: fractional settling rate (destructive metamorphism) at 0ᵒC for densities less than the critical density ρ_d (0.005 hr⁻¹)
```
[the remark this answers is not in the document]
Comment (a-charbon) [reply src/standalone/Snow/Snow.jl:71]: previous comment's resolution mitigates this comment (constants are stated in their original units for consistency)

File: ext/neural_snow/NeuralSnow.jl (hunk 29 above)
Comment (a-charbon) [reply ext/neural_snow/NeuralSnow.jl:133]: model() returns a 1-element Matrix, not the float itself. Comment has been added.

File: src/standalone/Snow/Snow.jl [hunk 35]
```diff
@@ -42,6 +42,49 @@ and is used as a bulk snow model.
...
+- c4: constant for adjusting fractional settling due to temperature (0.1 ᵒC⁻¹)
```
[the remark this answers is not in the document]
Comment (a-charbon) [reply src/standalone/Snow/Snow.jl:72]: see previous comments (maintain original units)

File: src/standalone/Snow/Snow.jl (hunk 2 above)
Comment (a-charbon) [reply src/standalone/Snow/Snow.jl:388]: Part of the NaN issues I messaged you about.

File: src/standalone/Snow/snow_parameterizations.jl (hunk 5 above)
Comment (a-charbon) [reply src/standalone/Snow/snow_parameterizations.jl:174]: I dont remember changing these, I'm guessing these are from some of your edits?

File: src/standalone/Snow/Snow.jl (hunk 6 above)
Comment (a-charbon) [reply src/standalone/Snow/Snow.jl:337]: This change was part of our consideration of the depth calculation when depth is directly available before making this change.

File: src/standalone/Snow/snow_parameterizations.jl (hunk 8 above)
Comment (a-charbon) [reply src/standalone/Snow/snow_parameterizations.jl:206]: Let's talk about this one.

File: src/standalone/Snow/snow_parameterizations.jl (hunk 10 above)
Comment (a-charbon) [reply src/standalone/Snow/snow_parameterizations.jl:471]: These are pulled directly from Snow17 for consistency.

File: src/standalone/Snow/snow_parameterizations.jl (hunk 11 above)
Comment (a-charbon) [reply src/standalone/Snow/snow_parameterizations.jl:483]: I can change the names to make this clearer, but these are explicitly implementing the paramterizations from snow_17 for use with the Anderson1976 density type (my intention was for them to _only_ be used with this type, and therefore preserve consisteny) They take input temperatures as celsius - I could change the units of the function and all other associated snow17 functions, but I already alternatively handle this in dzdt() for Anderson1976 models where I just convert the passed air temperature (in Kelvin) to celsius, and then pass it to the snow17 functions. Let me know which you prefer, as I doubt we'd want to use these parameterizations outside the Anderson1976 type.

If you'd prefer, i can just stick a comment above all these functions that says:
"#**************** BEGIN SNOW17 FUNCTIONS FOR Anderson1976 DENSITY TYPE ****************"
this is why all Anderson1976 functions were moved together at the bottom of the page.

File: src/standalone/Snow/snow_parameterizations.jl (hunk 12 above)
Comment (a-charbon) [reply src/standalone/Snow/snow_parameterizations.jl:473]: passed temperatures to this function are in celsius as of now (see above comment)

File: src/standalone/Snow/snow_parameterizations.jl (hunk 16 above)
Comment (a-charbon) [reply src/standalone/Snow/snow_parameterizations.jl:534]: I was mixing up computational order on auxillary vs prognostic variables and which values would get reported with the value at the time of saving, i've removed this now. (aka, if the snowpack is melted to z=SWE=0, in the saved values would we see z=SWE=0 and q_l = something < 1? happy to talk about this).

File: test/standalone/Snow/parameterizations.jl (hunk 17 above)
Comment (a-charbon) [reply test/standalone/Snow/parameterizations.jl:76]: This will be dependent on our discussion about the above.

File: src/integrated/soil_snow_model.jl (hunk 19 above)
Comment (a-charbon) [reply src/integrated/soil_snow_model.jl:256]: The two are identical for the given configuration, it only changes if the type of the density model changes.

File: src/integrated/soil_snow_model.jl (hunk 20 above)
Comment (a-charbon) [reply src/integrated/soil_snow_model.jl:398]: this was an artifact of the older state of my branch during the update, i believe

Comment (a-charbon) [reply src/integrated/soil_snow_model.jl:398]: artifact from rebasing i believe, this has been rectified.

File: src/integrated/soil_snow_model.jl (hunk 21 above)
Comment (a-charbon) [reply src/integrated/soil_snow_model.jl:401]: same as above.

File: ext/neural_snow/NeuralSnow.jl (hunk 23 above)
Comment (a-charbon) [reply ext/neural_snow/NeuralSnow.jl:17]: sounds good to me, i'll use that version.

**Comment by kmdeck** [comment]
superceded by #946
